planning/behavior_path_planner/src/scene_module/avoidance/avoidance_utils.cpp

```diff
@@ -264,6 +263,305 @@ tier4_autoware_utils::Polygon2d expandPolygon(
+    static_cast<size_t>(std::max(0, static_cast<int>(std::min(seg_idx1, seg_idx2)) - 5));
+  const size_t end_idx = static_cast<size_t>(std::min(
+    static_cast<int>(bound.size()) - 1, static_cast<int>(std::max(seg_idx1, seg_idx2)) + 1 + 5));
+  for (int i = start_idx; i < static_cast<int>(end_idx); ++i) {
+    const auto intersect_point = intersect(p1, p2, bound.at(i), bound.at(i + 1));
+    if (intersect_point) {
+      std::pair<size_t, geometry_msgs::msg::Point> result;
+      result.first = static_cast<size_t>(i);
+      result.second = intersect_point.get();
+      return result;
+    }
+  }
+  return boost::none;
+}
+
+PolygonPoint transformBoundFrenetCoordinate(
+  const std::vector<geometry_msgs::msg::Point> & points, const geometry_msgs::msg::Point & point)
+{
+  const size_t seg_idx = motion_utils::findNearestSegmentIndex(points, point);
+  const double lon_dist_to_segment =
+    motion_utils::calcLongitudinalOffsetToSegment(points, seg_idx, point);
+  const double lat_dist = motion_utils::calcLateralOffset(points, point, seg_idx);
+  return PolygonPoint{point, seg_idx, lon_dist_to_segment, lat_dist};
+}
+
+std::vector<PolygonPoint> generatePolygonInsideBounds(
+  const std::vector<Point> & bound, const std::vector<Point> & edge_points,
+  const bool is_object_right)
+{
+  std::vector<PolygonPoint> full_polygon;
+  for (size_t i = 0; i < edge_points.size(); ++i) {
+    const auto polygon_point = transformBoundFrenetCoordinate(bound, edge_points.at(i));
+    full_polygon.push_back(polygon_point);
+  }
+
+  std::vector<PolygonPoint> inside_poly;
+  for (int i = 0; i < static_cast<int>(full_polygon.size()); ++i) {
+    const auto & curr_poly = full_polygon.at(i);
+    const auto & prev_poly = full_polygon.at(i == 0 ? full_polygon.size() - 1 : i - 1);
+
+    const bool is_curr_outside = curr_poly.is_outside_bounds(is_object_right);
+    const bool is_prev_outside = prev_poly.is_outside_bounds(is_object_right);
+
+    if (is_curr_outside && is_prev_outside) {
+      continue;
+    }
+    if (!is_curr_outside && !is_prev_outside) {
+      inside_poly.push_back(curr_poly);
+      continue;
+    }
+
+    const auto intersection = intersectBound(
+      prev_poly.point, curr_poly.point, bound, prev_poly.bound_seg_idx, curr_poly.bound_seg_idx);
+    if (!intersection) {
+      continue;
+    }
+    const double lon_dist = motion_utils::calcLongitudinalOffsetToSegment(
+      bound, intersection->first, intersection->second);
+    const auto intersect_point =
+      PolygonPoint{intersection->second, intersection->first, lon_dist, 0.0};
+
+    if (is_prev_outside && !is_curr_outside) {
+      inside_poly.push_back(intersect_point);
+      inside_poly.push_back(curr_poly);
+      continue;
+    }
+    // Here is if (!is_prev_outside && is_curr_outside).
+    inside_poly.push_back(prev_poly);
+    inside_poly.push_back(intersect_point);
+    continue;
+  }
+
+  return inside_poly;
+}
+
+std::vector<geometry_msgs::msg::Point> convertToPoints(
```

Comment:
I would prefer `convertToGeometryPoints` for better clarity

Reply:
I agree.
Fixed https://github.com/autowarefoundation/autoware.universe/pull/2880/commits/2f6b217d9f8a572831f5e8b36134f1ce9f4a6be3